File: tools/render-pdf.py
#!/usr/bin/env python3
"""Render an HTML documentation file to PDF using WeasyPrint.

Replaces `electron-pdf --marginsType 0 --input X --output Y`.

Usage: render-pdf.py <input.html> <output.pdf>

WeasyPrint is a CSS Paged Media engine (no browser/Chromium). Page size and
margins come from the @page rule in the HTML; page-break markers in the source
(`<div style="page-break-after: always"></div>`) are honored natively. Relative
asset paths (images, etc.) resolve against the input file's directory.
"""

import sys
from pathlib import Path

from weasyprint import HTML


def main() -> int:
    if len(sys.argv) != 3:
        print("Usage: render-pdf.py <input.html> <output.pdf>", file=sys.stderr)
        return 1

    input_path = Path(sys.argv[1]).resolve()
    output_path = Path(sys.argv[2]).resolve()

    HTML(filename=str(input_path)).write_pdf(str(output_path))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/render-pdf.mjs
#!/usr/bin/env node
// Render an HTML file to PDF via headless Chromium (puppeteer).
// Drop-in replacement for `electron-pdf --marginsType 0 --input X --output Y`.
//
// Usage: node tools/render-pdf.mjs <input.html> <output.pdf>

import { pathToFileURL } from "node:url";
import { resolve } from "node:path";

const [, , input, output] = process.argv;
if (!input || !output) {
  console.error("Usage: render-pdf.mjs <input.html> <output.pdf>");
  process.exit(1);
}

const puppeteer = (await import("puppeteer")).default;

const browser = await puppeteer.launch({
  headless: true,
  args: ["--no-sandbox", "--disable-setuid-sandbox"],
});
try {
  const page = await browser.newPage();
  await page.emulateMediaType("print");
  await page.goto(pathToFileURL(resolve(input)).href, {
    waitUntil: "networkidle0",
  });
  // electron-pdf --marginsType 0 is "default" margins (~0.4in / 10mm), NOT zero.
  // Match those to keep our output visually consistent with the previous pipeline.
  await page.pdf({
    path: resolve(output),
    format: "Letter",
    printBackground: true,
    margin: {
      top: "0.4in",
      right: "0.4in",
      bottom: "0.4in",
      left: "0.4in",
    },
  });
} finally {
  await browser.close();
}
